src/main/java/com/aws/greengrass/lifecyclemanager/KernelLifecycle.java

```diff
@@ -592,4 +594,31 @@ public void shutdown(int timeoutSeconds) {
     GreengrassService getMain() {
         return mainService;
     }
+
+    /**
+     * Check if all services has reached to terminal state: RUNNING, FINISHED or BROKEN.
+     *
+     */
+    public boolean allServicesInTerminalState() {
+        //1. Check all other component status
+        List<GreengrassService> servicesToTrack = kernel.orderedDependencies().stream()
+                .filter(GreengrassService::shouldAutoStart).collect(Collectors.toList());
+        boolean allReachedTerminalState = true;
+        for (GreengrassService service : servicesToTrack) {
+            State state = service.getState();
+            if (!service.reachedDesiredState()) {
+                allReachedTerminalState = false;
+                break;
+            }
+            if (State.RUNNING.equals(state) || State.FINISHED.equals(state) || !service.shouldAutoStart()
```

Comment:
`!service.shouldAutoStart()`
this is always false, right?

Reply:
Yes, but if we look at the loop from `waitForServicesToStart` it's also checking this condition. I think the meaning for it to be here is in future, if we have some components should not auto start but in the desired state, we also want to consider that service is running

Reply:
Sure, but you are already filtering out services with should auto start as true at 604. If your filtered list of services will always autostart then `!service.shouldAutoStart()` will always be false, right?

Reply:
Right, updated